In [75]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt 
import numpy as np
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import mean_absolute_error

In [2]:
training_data =  pd.read_csv("HR_comma_sep.csv")
training_data.describe()

satisfaction_level  last_evaluation  number_project  \
count        14999.000000     14999.000000    14999.000000   
mean             0.612834         0.716102        3.803054   
std              0.248631         0.171169        1.232592   
min              0.090000         0.360000        2.000000   
25%              0.440000         0.560000        3.000000   
50%              0.640000         0.720000        4.000000   
75%              0.820000         0.870000        5.000000   
max              1.000000         1.000000        7.000000   

       average_montly_hours  time_spend_company  Work_accident          left  \
count          14999.000000        14999.000000   14999.000000  14999.000000   
mean             201.050337            3.498233       0.144610      0.238083   
std               49.943099            1.460136       0.351719      0.425924   
min               96.000000            2.000000       0.000000      0.000000   
25%              156.000000            3.000000       0.000000      0.000000   
50%              200.000000            3.000000       0.000000      0.000000   
75%              245.000000            4.000000       0.000000      0.000000   
max              310.000000           10.000000       1.000000      1.000000   

       promotion_last_5years  
count           14999.000000  
mean                0.021268  
std                 0.144281  
min                 0.000000  
25%                 0.000000  
50%                 0.000000  
75%                 0.000000  
max                 1.000000

In [66]:
training_features = ["satisfaction_level", "last_evaluation", 
                        "number_project", "average_montly_hours",
                            "time_spend_company", "Work_accident", "promotion_last_5years"]

X_train, X_test, y_train, y_test = train_test_split(training_data[training_features], 
                                                        training_data[["left"]],
                                                            random_state=50, train_size=0.7)

X_train.describe()

satisfaction_level  last_evaluation  number_project  \
count        10499.000000     10499.000000    10499.000000   
mean             0.616350         0.717691        3.813792   
std              0.249297         0.170796        1.231801   
min              0.090000         0.360000        2.000000   
25%              0.440000         0.560000        3.000000   
50%              0.650000         0.720000        4.000000   
75%              0.820000         0.870000        5.000000   
max              1.000000         1.000000        7.000000   

       average_montly_hours  time_spend_company  Work_accident  \
count          10499.000000        10499.000000   10499.000000   
mean             201.477284            3.511192       0.144490   
std               49.757937            1.476540       0.351603   
min               96.000000            2.000000       0.000000   
25%              156.000000            3.000000       0.000000   
50%              201.000000            3.000000       0.000000   
75%              245.000000            4.000000       0.000000   
max              310.000000           10.000000       1.000000   

       promotion_last_5years  
count           10499.000000  
mean                0.021335  
std                 0.144507  
min                 0.000000  
25%                 0.000000  
50%                 0.000000  
75%                 0.000000  
max                 1.000000

In [67]:
model = Sequential(
        [Dense(input_dim = X_train.shape[1],units=100,activation="relu"),
         Dense(units=25,activation="relu"),
         Dense(units=10,activation="relu"),
         Dense(units=1, activation="sigmoid")
        ], name = "my_model")

model.summary()

Model: "my_model"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_88 (Dense)            (None, 100)               800       
                                                                 
 dense_89 (Dense)            (None, 25)                2525      
                                                                 
 dense_90 (Dense)            (None, 10)                260       
                                                                 
 dense_91 (Dense)            (None, 1)                 11        
                                                                 
Total params: 3,596
Trainable params: 3,596
Non-trainable params: 0
_________________________________________________________________


In [73]:
model.compile(optimizer = 'adam', loss="binary_crossentropy")
model.fit(X_train,y_train,epochs=1000)

Epoch 1/1000
329/329 [==============================] - 4s 4ms/step - loss: 0.1679
Epoch 2/1000
329/329 [==============================] - 1s 3ms/step - loss: 0.1692
Epoch 3/1000
329/329 [==============================] - 1s 3ms/step - loss: 0.1705
Epoch 4/1000
329/329 [==============================] - 1s 3ms/step - loss: 0.1684
Epoch 5/1000
329/329 [==============================] - 2s 5ms/step - loss: 0.1574
Epoch 6/1000
329/329 [==============================] - 2s 5ms/step - loss: 0.1651
Epoch 7/1000
329/329 [==============================] - 2s 5ms/step - loss: 0.1592
Epoch 8/1000
329/329 [==============================] - 2s 5ms/step - loss: 0.1624
Epoch 9/1000
329/329 [==============================] - 1s 4ms/step - loss: 0.1601
Epoch 10/1000
329/329 [==============================] - 1s 4ms/step - loss: 0.1604
Epoch 11/1000
329/329 [==============================] - 1s 4ms/step - loss: 0.1516
Epoch 12/1000
329/329 [==============================] - 1s 4ms/step - loss: 0.1556
E

In [77]:
predictions = model.predict(X_test)

threshold = 0.5

predictions = np.where(predictions < threshold, np.floor(predictions), np.ceil(predictions))
print(1 - mean_absolute_error(predictions, y_test))

141/141 [==============================] - 1s 4ms/step
0.9704444444444444


In [78]:
model.evaluate(X_test, y_test)

141/141 [==============================] - 1s 2ms/step - loss: 0.1099


0.10990817844867706

In [84]:
model = LogisticRegression()
model.fit(X_train, y_train)
predictions = model.predict(X_test)
model.score(X_test,y_test)

c:\Users\olaoy\AppData\Local\Programs\Python\Python310\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\olaoy\AppData\Local\Programs\Python\Python310\lib\site-packages\sklearn\linear_model\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


0.7655555555555555

In [125]:
z = lambda x,w,b: np.dot(w,x) + b

In [126]:
g =  lambda z: 1 / (1 + np.exp(-z))


In [137]:
def compute_gradient(X,y,w,b):
    dj_dw = np.zeros(w.shape)
    dj_db = 0.
    m = X.shape[0]
    X, y = X.values, y.values

    for i in range(m):
        loss = g(z(X[i],w,b)) - y[i]
        dj_dw += loss * X[i]
        dj_db += loss[0]

    return dj_dw / m , dj_db / m

In [138]:
def gradient_descent(X,y,w,b,nums_iter,alpha):
    for i in range(nums_iter):
        dj_dw, dj_db = compute_gradient(X,y,w,b)
        w -= alpha * dj_dw
        b -= alpha * dj_db

    return w,b

In [149]:
w_,b_ = gradient_descent(X_train,y_train,np.zeros(7),0.,1000,0.01)


In [150]:
def predict(X):
    m = X.shape[0]
    predictions = np.zeros(m)
    X = X.values

    for i in range(m):
        res = g(z(X[i], w_, b_))
        predictions[i] = np.ceil(res) if res >= 0.5 else np.floor(res)

    return predictions


y_pred = predict(X_test)
print(1 - mean_absolute_error(y_test, y_pred))

0.762


ANN came out top with 97% accuracy, Sklearn's model came second with 76.5% accuracy and mine last with 76.2% accuracy. I would tweak the num_iters for gradient one more time. 